File: app/api/aica_api/routers/run_plans.py
"""Run-plans router — /api/run-plans."""

from __future__ import annotations

import os
from datetime import datetime, timezone
from typing import Any

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

from aica_api.config import settings
from aica_api.services.package_registry import PackageRegistry
from aica_api.services.run_plan import create_draft, get_draft_entry, regenerate_draft
from aica_api.services.scenario_registry import ScenarioRegistry

router = APIRouter()


# ── Plan-ID generation (only place timestamps/randomness allowed) ─────────────


def _make_plan_id() -> str:
    """Generate a unique plan_id: plan_<YYYYMMDD-HHMMSS>_<6-hex>."""
    ts = datetime.now(timezone.utc).strftime("%Y%m%d-%H%M%S")
    rand = os.urandom(3).hex()
    return f"plan_{ts}_{rand}"


# ── Request body models ───────────────────────────────────────────────────────


class CreateRunPlanBody(BaseModel):
    package_id: str
    scenario_id: str
    route_facts: Any = None  # optional; computed locally if absent
    presets: dict = {}
    profiles: Any = None
    parameters: dict = {}
    hyperparameters: dict = {}
    run_mode: str = "standard"


class RegenerateRunPlanBody(BaseModel):
    presets: dict = {}
    parameters: dict = {}
    hyperparameters: dict = {}


# ── Endpoints ─────────────────────────────────────────────────────────────────


@router.post("/api/run-plans", status_code=201)
def create_run_plan_endpoint(body: CreateRunPlanBody):
    """Validate edits, build draft event plan, register draft. 400 on errors."""
    pkg_reg = PackageRegistry(settings.packages_dir)
    sc_reg = ScenarioRegistry(settings.scenarios_dir)

    package = pkg_reg.get(body.package_id)
    if package is None:
        raise HTTPException(
            status_code=400,
            detail=f"Package {body.package_id!r} not found or invalid",
        )

    scenario = sc_reg.get(body.scenario_id)
    if scenario is None:
        raise HTTPException(
            status_code=400,
            detail=f"Scenario {body.scenario_id!r} not found or invalid",
        )

    if not pkg_reg.is_compatible(package, scenario):
        raise HTTPException(
            status_code=400,
            detail=(
                f"Package {body.package_id!r} is not compatible with "
                f"scenario {body.scenario_id!r} (type={scenario.type!r})"
            ),
        )

    plan_id = _make_plan_id()
    draft = create_draft(
        plan_id=plan_id,
        package=package,
        scenario=scenario,
        presets=body.presets,
        parameters=body.parameters,
        hyperparameters=body.hyperparameters,
        run_mode=body.run_mode,
    )

    if draft.validation_errors:
        raise HTTPException(
            status_code=400,
            detail={
                "detail": "One or more parameter/hyperparameter values are invalid.",
                "validation_errors": draft.validation_errors,
            },
        )

    return {
        "plan_id": draft.plan_id,
        "draft_plan": draft.draft_event_plan,
        "effective_setup": draft.effective_setup,
        "validation_errors": draft.validation_errors,
    }


@router.post("/api/run-plans/{plan_id}/regenerate")
def regenerate_run_plan_endpoint(plan_id: str, body: RegenerateRunPlanBody):
    """Regenerate draft with updated presets/parameters/hyperparameters.

    404 for unknown plan, 400 for validation errors.
    """
    if get_draft_entry(plan_id) is None:
        raise HTTPException(
            status_code=404,
            detail=f"Run plan {plan_id!r} not found",
        )

    try:
        draft = regenerate_draft(
            plan_id=plan_id,
            presets=body.presets,
            parameters=body.parameters,
            hyperparameters=body.hyperparameters,
        )
    except ValueError as exc:
        raise HTTPException(status_code=404, detail=str(exc))

    if draft.validation_errors:
        raise HTTPException(
            status_code=400,
            detail={
                "detail": "One or more parameter/hyperparameter values are invalid.",
                "validation_errors": draft.validation_errors,
            },
        )

    return {
        "plan_id": draft.plan_id,
        "draft_plan": draft.draft_event_plan,
        "effective_setup": draft.effective_setup,
        "validation_errors": draft.validation_errors,
    }

File: htmlapp/frontend/src/engine/worker/handlers/run_plans.ts
import type {
  PackageManifest,
  ScenarioDef,
  SetupValue,
  RunPlanResponse,
  RouteFacts,
  DisplayRoute,
  ProfileOverrides,
  ValidationError,
} from '../../../api/types'
import { RunPlanError } from '../../../api/types'
import { packageRegistry } from '../../services/package_registry'
import { scenarioRegistry } from '../../services/scenario_registry'
import {
  createDraft,
  regenerateDraft,
  getDraftEntry,
  type PackageManifestM2,
} from '../../run_plan'
import type { ScenarioDefM2 } from '../../event_plan'

/** Loose stand-in for Python's `!r` repr formatting used in router error messages. */
function pyReprValue(value: unknown): string {
  if (typeof value === 'string') return `'${value}'`
  return String(value)
}

// plan_id generation: monotonic in-memory counter for determinism.
let _planIdCounter = 0
export function makePlanId(): string {
  _planIdCounter += 1
  return `plan_${String(_planIdCounter).padStart(6, '0')}`
}

const VALID_INITIAL_STATE_KEYS = ['drowsiness_level', 'fatigue_level']
const VALID_CONTEXT_KEYS = ['child_passenger', 'familiar_route']

/** Mirrors the router's initial_state key/range validation. */
function validateInitialStateBody(initialState: Record<string, unknown>): ValidationError[] {
  const errors: ValidationError[] = []
  for (const [key, value] of Object.entries(initialState)) {
    if (!VALID_INITIAL_STATE_KEYS.includes(key)) {
      errors.push({
        field: `initial_state.${key}`,
        message: `Unknown initial_state key ${pyReprValue(key)}. Valid keys: ['drowsiness_level', 'fatigue_level']`,
      })
    } else if (typeof value !== 'number') {
      errors.push({
        field: `initial_state.${key}`,
        message: `initial_state.${key} must be a number in [0, 100]; got ${pyReprValue(value)}`,
      })
    } else if (!(value >= 0 && value <= 100)) {
      errors.push({
        field: `initial_state.${key}`,
        message: `initial_state.${key} must be in [0, 100]; got ${pyReprValue(value)}`,
      })
    }
  }
  return errors
}

/** Mirrors the router's context_overrides key/type validation. */
function validateContextOverridesBody(contextOverrides: Record<string, unknown>): ValidationError[] {
  const errors: ValidationError[] = []
  for (const [key, value] of Object.entries(contextOverrides)) {
    if (!VALID_CONTEXT_KEYS.includes(key)) {
      errors.push({
        field: `context_overrides.${key}`,
        message: `Unknown context key ${pyReprValue(key)}. Valid keys: ['child_passenger', 'familiar_route']`,
      })
    } else if (typeof value !== 'boolean') {
      errors.push({
        field: `context_overrides.${key}`,
        message: `context_overrides.${key} must be a boolean; got ${pyReprValue(value)}`,
      })
    }
  }
  return errors
}

export async function runPlansCreate(params: {
  packageId: string
  scenarioId: string
  parameters?: Record<string, SetupValue>
  hyperparameters?: Record<string, SetupValue>
  presets?: Record<string, unknown>
  runMode?: string
  routeId?: string
  routeSource?: string
  routeFacts?: RouteFacts | null
  displayRoute?: DisplayRoute | null
  profiles?: ProfileOverrides | null
  initialState?: { drowsiness_level?: number; fatigue_level?: number }
  contextOverrides?: { child_passenger?: boolean; familiar_route?: boolean }
}): Promise<RunPlanResponse> {
  let pkgManifest: PackageManifest
  try {
    pkgManifest = await packageRegistry.get(params.packageId)
  } catch {
    throw new RunPlanError(`Package ${pyReprValue(params.packageId)} not found or invalid`)
  }

  let scenario: ScenarioDef
  try {
    scenario = await scenarioRegistry.get(params.scenarioId)
  } catch {
    throw new RunPlanError(`Scenario ${pyReprValue(params.scenarioId)} not found or invalid`)
  }

  if (!packageRegistry.isCompatible(pkgManifest, scenario)) {
    throw new RunPlanError(
      `Package ${pyReprValue(params.packageId)} is not compatible with scenario ${pyReprValue(params.scenarioId)} (type=${pyReprValue(scenario.type)})`,
    )
  }

  const routeSource = params.routeSource ?? 'local'
  if (routeSource === 'maps') {
    if (params.routeId == null) {
      throw new RunPlanError(
        "route_id is required when route_source is 'maps'. Pass the route_id from the chosen /api/routes/analyze alternative.",
      )
    }
    if (params.routeFacts == null) {
      throw new RunPlanError(
        "route_facts is required when route_source is 'maps'. A run cannot be created without route facts. " +
          'Pass the route_facts from the chosen /api/routes/analyze alternative.',
      )
    }
  }

  const routeFacts = routeSource === 'maps' ? (params.routeFacts ?? null) : null
  const displayRoute = routeSource === 'maps' ? (params.displayRoute ?? null) : null

  const profilesDict: Record<string, unknown> | null = params.profiles ?? null

  if (params.initialState != null) {
    const initErrors = validateInitialStateBody(params.initialState)
    if (initErrors.length > 0) {
      throw new RunPlanError('One or more initial_state values are invalid.', initErrors)
    }
  }

  if (params.contextOverrides != null) {
    const ctxErrors = validateContextOverridesBody(params.contextOverrides)
    if (ctxErrors.length > 0) {
      throw new RunPlanError('One or more context_overrides values are invalid.', ctxErrors)
    }
  }

  const planId = makePlanId()
  const { draft } = createDraft({
    planId,
    package: pkgManifest as unknown as PackageManifestM2,
    scenario: scenario as unknown as ScenarioDefM2,
    presets: params.presets ?? {},
    parameters: params.parameters ?? {},
    hyperparameters: params.hyperparameters ?? {},
    runMode: params.runMode ?? 'standard',
    routeFacts,
    routeSource: routeSource as 'maps' | 'local',
    displayRoute,
    profiles: profilesDict,
    initialState: params.initialState ?? null,
    contextOverrides: params.contextOverrides ?? null,
  })

  if (draft.validation_errors.length > 0) {
    throw new RunPlanError('One or more parameter/hyperparameter values are invalid.', draft.validation_errors)
  }

  return {
    plan_id: draft.plan_id,
    draft_plan: draft.draft_event_plan,
    effective_setup: draft.effective_setup,
    validation_errors: draft.validation_errors,
  }
}

export async function runPlansRegenerate(params: {
  planId: string
  patch: {
    parameters?: Record<string, SetupValue>
    hyperparameters?: Record<string, SetupValue>
    presets?: Record<string, unknown>
  }
}): Promise<RunPlanResponse> {
  if (getDraftEntry(params.planId) === null) {
    throw new RunPlanError(`Run plan ${pyReprValue(params.planId)} not found`)
  }

  let draft: ReturnType<typeof regenerateDraft>
  try {
    draft = regenerateDraft(params.planId, {
      presets: params.patch.presets ?? {},
      parameters: params.patch.parameters ?? {},
      hyperparameters: params.patch.hyperparameters ?? {},
    })
  } catch (exc) {
    throw new RunPlanError(exc instanceof Error ? exc.message : String(exc))
  }

  if (draft.validation_errors.length > 0) {
    throw new RunPlanError('One or more parameter/hyperparameter values are invalid.', draft.validation_errors)
  }

  return {
    plan_id: draft.plan_id,
    draft_plan: draft.draft_event_plan,
    effective_setup: draft.effective_setup,
    validation_errors: draft.validation_errors,
  }
}
